### PR TITLE
Use immutable secrets in `ManagedResource` library

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:
A successor of https://github.com/gardener/gardener/pull/7959

This PR aims to enhance the managed resource reconciliation flow. Right now if we change the underlying secret of a managedresource and we call `WaitUntilHealthy` the function might immediately return since all conditions are met even though the later function calls `CheckManagedResourceApplied`. `CheckManagedResourceApplied` checks the generation of the managedresource which does not change when an underlying secret is updated. This PR aims to solve this by using the managed resources combined with immutable secrets. This way the generation of the MR will be increased if the `secretRefs` change. Another goal of this PR is to provide seamless transition to users of the mr utils package.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
~There are still a lot of failing tests, but I wanted opinion on the work done until now. Please take a look @timuthy @rfranzke Most of the logic is in commits until https://github.com/gardener/gardener/commit/f644783428e513f6fc794989aabc5cdba68f178f
Release notes will be added once the PR is not a work in progress.~

**Release note**:

```breaking developer
Package `pkg/utils/managedresources` now works with immutable secrets for managed resources under the hood. Existing secrets will be marked for garbage collection and replaced with immutable ones during the first reconciliation of the managed resource.
```

```breaking developer
The `Secrets` type as well as the `Delete` functions for secrets were removed from `pkg/utils/managedresources/builder` since their usage was prone to errors. The higher level package `pkg/utils/managedresources` should be used instead.
```

```feature operator
The garbage collection controller now also considers managed resources when deciding if secrets/configmaps should be garbage collected.
```

### CODE REVIEW

File: pkg/utils/managedresources/builder/secrets.go [hunk 1]
```diff
@@ -65,6 +68,13 @@ func (s *Secret) WithKeyValues(keyValues map[string][]byte) *Secret {
 	return s
 }
 
+// Unique makes the secret unique and immutable. Returns the new and unique name of the secret and the builder object.
+// This function should be called at last place, i.e. after the secret object is constructed.
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/secrets.go]:
```suggestion
// This function should be called after the name and data of the secret was set.
```

File: pkg/utils/managedresources/builder/managedresources.go [hunk 2]
```diff
@@ -112,28 +115,94 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
...
+func markSecretsAsGarbageCollectable(ctx context.Context, c client.Client, secrets []*corev1.Secret) error {
+	for _, secret := range secrets {
+		if err := c.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+			return err
+		}
+		patch := client.StrategicMergeFrom(secret.DeepCopy())
+		if secret.Labels == nil {
+			secret.Labels = map[string]string{}
+		}
+		secret.Labels[references.LabelKeyGarbageCollectable] = references.LabelValueGarbageCollectable
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/managedresources.go]:
```suggestion
metav1.SetMetadataLabel(&secret.ObjectMeta, references.LabelKeyGarbageCollectable, references.LabelValueGarbageCollectable)
```

File: pkg/utils/managedresources/builder/managedresources.go [hunk 3]
```diff
@@ -112,28 +115,94 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
...
+			if err := markSecretsAsGarbageCollectable(ctx, m.client, secretsFromRefs(resource)); err != nil {
+				return err
+			}
+
+			return m.client.Create(ctx, resource)
+		}
+		return err
+	}
+
+	// mark all old secrets as garbage collectable
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/managedresources.go]: Can we add a TODO that this code is only relevant for migration and can be removed after a few releases, e.g. `v1.80`?

File: pkg/utils/managedresources/builder/resourcemanager_test.go [hunk 4]
```diff
@@ -229,7 +263,67 @@ var _ = Describe("Resource Manager", func() {
...
+				Expect(fakeClient.Create(ctx, s)).To(Succeed())
+			}
+
+			Expect(
+				NewManagedResource(fakeClient).
+					WithNamespacedName(namespace, name).
+					WithLabels(labels).
+					WithAnnotations(annotations).
+					WithSecretRef(secret1.Name).
+					WithSecretRef(secret2.Name).
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/resourcemanager_test.go:352]: The test says "should label existing managed resource secrets" but in reality `WithSecretRef` would be called with the new, immutable secrets (i.e. name changes). Can you please adapt the test to really ensure, that the "old" secrets are marked as garbage collectable?

File: pkg/utils/managedresources/builder/managedresources.go [hunk 5]
```diff
@@ -112,28 +115,94 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
...
 		}
 
-		resource.Spec = m.resource.Spec
+		obj.Spec = m.resource.Spec
+
+		// the annotations should be injected after the spec is updated!
+		utilruntime.Must(references.InjectAnnotations(obj))
+	}
+
+	if err := m.client.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil {
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/managedresources.go:141]: Can't we stick to `controllerutil.CreateOrUpdate` here?

File: pkg/utils/managedresources/managedresources.go [hunk 6]
```diff
@@ -186,20 +189,43 @@ func deployManagedResource(ctx context.Context, secret *builder.Secret, managedR
 	return nil
 }
 
-// Delete deletes the managed resource and its secret with the given name in the given namespace.
-func Delete(ctx context.Context, client client.Client, namespace string, name string, secretNameWithPrefix bool) error {
+// Delete deletes the managed resource and its secrets with the given name in the given namespace.
+func Delete(ctx context.Context, c client.Client, namespace string, name string, secretNameWithPrefix bool) error {
 	secretName := SecretName(name, secretNameWithPrefix)
```
Comment (timuthy) [inline pkg/utils/managedresources/managedresources.go]: This `secretName` is only required to deleted "old" secrets, isn't it? Once all secrets were migrated, we can probably drop this code as well. Can you add a TODO for that?

File: pkg/utils/managedresources/managedresources.go [hunk 7]
```diff
@@ -186,20 +189,43 @@ func deployManagedResource(ctx context.Context, secret *builder.Secret, managedR
...
+		if err := client.IgnoreNotFound(c.Delete(ctx, secret)); err != nil {
+			return fmt.Errorf("could not delete secret '%s' of managed resource: %w", client.ObjectKeyFromObject(secret).String(), err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("could not get managed resource '%s': %w", mrKey.String(), err)
 	}
 
-	if err := builder.NewSecret(client).
-		WithNamespacedName(namespace, secretName).
-		Delete(ctx); err != nil {
```
Comment (timuthy) [inline pkg/utils/managedresources/managedresources.go:201]: Should we drop the `Delete` function since it's rather "dangerous" to rely on?

**Review by timuthy — COMMENTED** [review]
Together with @ary1992 and @shafeeqes we had a first look at the changes, until commit 08fb2b076d92353776f59fc4f386d95dadcce583.

File: pkg/utils/managedresources/builder/managedresources.go (hunk 5 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/builder/managedresources.go:141]: We probably can, but I do not find it intuitive to mutate secrets in the `controllerutil.CreateOrUpdate` function called on a managed resource object.

File: pkg/utils/managedresources/managedresources.go (hunk 7 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/managedresources.go:201]: Do you mean to drop it from the `builder/secrets.go` file? I am OK with that suggestion since its usage is error prone.

File: pkg/utils/managedresources/builder/resourcemanager_test.go (hunk 4 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/builder/resourcemanager_test.go:352]: Done

**Comment by dimityrmirchev** [comment]
@timuthy, @ary1992, @shafeeqes I addressed the initial comments. PTAL

**Comment by timuthy** [comment]
@dimityrmirchev thanks for addressing the review comments. I haven't managed to check the PR in detail again but will do ASAP. Thanks for being patient 🙏

File: pkg/utils/managedresources/managedresources.go [hunk 8]
```diff
@@ -195,6 +195,7 @@ func deployManagedResource(ctx context.Context, secret *builder.Secret, managedR
 
 // Delete deletes the managed resource and its secrets with the given name in the given namespace.
 func Delete(ctx context.Context, c client.Client, namespace string, name string, secretNameWithPrefix bool) error {
+	// TODO(dimityrmirchev): This can be dropped once old secrets are marked as garbage collectable
```
Comment (shafeeqes) [inline pkg/utils/managedresources/managedresources.go]: Probably add `v1.90` here too?

File: pkg/utils/managedresources/builder/managedresources.go [hunk 9]
```diff
@@ -152,6 +152,9 @@ func (m *ManagedResource) Reconcile(ctx context.Context) error {
 	}
 
 	// mark all old secrets as garbage collectable
+	// TODO(dimityrmirchev): We can remove this in a future release since this is migration code
+	// Since the removal of this code can cause leaks of managed resource secrets
+	// we want to give a greater time period than usual, i.e. v1.90 of Gardener
```
Comment (shafeeqes) [inline pkg/utils/managedresources/builder/managedresources.go]: Nit: We can just mention `Remove this in v1.90`

File: pkg/utils/managedresources/builder/managedresources.go [hunk 10]
```diff
@@ -142,7 +142,7 @@ func (m *ManagedResource) Reconcile(ctx context.Context) error {
 			mutateFn(resource)
 
 			// only mark the new secrets
-			if err := markSecretsAsGarbageCollectable(ctx, m.client, secretsFromRefs(resource)); err != nil {
+			if err := markSecretsAsGarbageCollectable(ctx, m.client, secretsFromRefs(resource, map[string]struct{}{})); err != nil {
```
Comment (shafeeqes) [inline pkg/utils/managedresources/builder/managedresources.go]: Why not use `sets.Set[string]`?

File: docs/concepts/resource-manager.md [hunk 11]
```diff
@@ -252,7 +252,7 @@ The purpose of this controller is cleaning up such immutable `ConfigMap`s/`Secre
 The following algorithm is implemented in the GC controller:
 
 1. List all `ConfigMap`s and `Secret`s labeled with `resources.gardener.cloud/garbage-collectable-reference=true`.
-1. List all `Deployment`s, `StatefulSet`s, `DaemonSet`s, `Job`s, `CronJob`s, `Pod`s and for each of them:
+1. List all `Deployment`s, `StatefulSet`s, `DaemonSet`s, `Job`s, `CronJob`s, `Pod`s, `ManagedResource`s and for each of them:
```
Comment (shafeeqes) [inline docs/concepts/resource-manager.md:255]: Maybe mention listing `ManagedResource` is only when it's configured?

File: pkg/resourcemanager/controller/garbagecollector/reconciler.go [hunk 12]
```diff
@@ -99,6 +100,10 @@ func (r *Reconciler) Reconcile(reconcileCtx context.Context, _ reconcile.Request
 		}
 	)
 
+	if r.Config.ConsiderManagedResourceSecrets != nil && *r.Config.ConsiderManagedResourceSecrets {
```
Comment (shafeeqes) [inline pkg/resourcemanager/controller/garbagecollector/reconciler.go]:
```suggestion
	if pointer.BoolDeref(r.Config.ConsiderManagedResourceSecrets, false){
```

File: pkg/resourcemanager/apis/config/types.go [hunk 13]
```diff
@@ -144,6 +144,9 @@ type GarbageCollectorControllerConfig struct {
 	Enabled bool
 	// SyncPeriod is the duration how often the controller performs its reconciliation.
 	SyncPeriod *metav1.Duration
+	// ConsiderManagedResourceSecrets specifies if managed resources should be considered
+	// as owners of a garbage collectable secret.
+	ConsiderManagedResourceSecrets *bool
```
Comment (shafeeqes) [inline pkg/resourcemanager/apis/config/types.go]: Also update https://github.com/gardener/gardener/blob/904f0ddef5d87a93d0838b9eefe99051c583d35d/charts/gardener/resource-manager/values.yaml#L76-L78

Comment (shafeeqes) [inline pkg/resourcemanager/apis/config/types.go]: Isn't `ConsiderManagedResourceSecrets` name a bit confusing? Maybe just `ConsiderManagedResources`?

File: pkg/operation/botanist/operatingsystemconfig_test.go [hunk 14]
```diff
@@ -323,9 +324,11 @@ var _ = Describe("operatingsystemconfig", func() {
 
 						// managed resource secret reconciliation for executor scripts for worker pools
 						// worker pool 1
+						// worker1ExecutorScript, _ := ExecutorScriptFn([]byte(worker1OriginalContent), cloudConfigExecutionMaxDelaySeconds, hyperkubeImage.ToImage(&kubernetesVersion), kubernetesVersion, nil, worker1OriginalCommand, worker1OriginalUnits, worker1OriginalFiles)
+						// kubernetesClientSeed.EXPECT().Get(ctx, kubernetesutils.Key(namespace, "managedresource-shoot-cloud-config-execution-"+worker1Name), gomock.AssignableToTypeOf(&corev1.Secret{}))
+						// kubernetesClientSeed.EXPECT().Update(ctx, &corev1.Secret{
```
Comment (shafeeqes) [inline pkg/operation/botanist/operatingsystemconfig_test.go]: Leftover from testing?

File: pkg/component/istio/istio_test.go [hunk 15]
```diff
@@ -2325,7 +2339,7 @@ spec:
 			It("should successfully deploy all resources", func() {
 				Expect(c.Get(ctx, client.ObjectKeyFromObject(managedResourceIstioSecret), managedResourceIstioSecret)).To(Succeed())
 				Expect(managedResourceIstioSecret.Type).To(Equal(corev1.SecretTypeOpaque))
-				Expect(managedResourceIstioSecret.Data).To(HaveLen(15))
+				Expect(managedResourceIstioSecret.Data).To(HaveLen(13))
```
Comment (shafeeqes) [inline pkg/component/istio/istio_test.go:2342]: Why did the length change now?

File: pkg/utils/managedresources/builder/secrets.go [hunk 16]
```diff
@@ -50,16 +51,41 @@ func (s *Secret) WithNamespacedName(namespace, name string) *Secret {
...
+// if it already exists in the current labels.
 func (s *Secret) WithLabels(labels map[string]string) *Secret {
-	s.secret.Labels = labels
+	if s.secret.Labels == nil {
+		s.secret.Labels = cloneLabels(labels)
+		return s
+	}
+
+	_, ok := s.secret.Labels[references.LabelKeyGarbageCollectable]
+	if ok && s.secret.Immutable != nil && *s.secret.Immutable {
```
Comment (shafeeqes) [inline pkg/utils/managedresources/builder/secrets.go]: You could use `pointer.BoolDeref`

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go [hunk 17]
```diff
@@ -326,25 +327,27 @@ func (r *Reconciler) delete(
...
 
-	secret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      controllerInstallation.Name,
-			Namespace: v1beta1constants.GardenNamespace,
-		},
-	}
-	if err := r.SeedClientSet.Client().Delete(seedCtx, secret); client.IgnoreNotFound(err) != nil {
-		conditionInstalled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionInstalled, gardencorev1beta1.ConditionFalse, "DeletionFailed", fmt.Sprintf("Deletion of ManagedResource secret %q failed: %+v", controllerInstallation.Name, err))
+	fiveSecond := time.Second * 5
```
Comment (shafeeqes) [inline pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: Maybe define a variable for this like `RequeueDurationWhenResourceDeletionStillPresent ` and overwrite it in test?

**Review by shafeeqes — COMMENTED** [review]
I have some more commits left to review, will make it by tomorrow.

File: hack/test-integration.env [hunk 18]
```diff
@@ -1,7 +1,7 @@
 # reduce flakiness in contended pipelines
 export KUBEBUILDER_CONTROLPLANE_START_TIMEOUT=2m
-export GOMEGA_DEFAULT_EVENTUALLY_TIMEOUT=5s
-export GOMEGA_DEFAULT_EVENTUALLY_POLLING_INTERVAL=200ms
+export GOMEGA_DEFAULT_EVENTUALLY_TIMEOUT=10s
```
Comment (shafeeqes) [inline hack/test-integration.env]: Please follow https://github.com/gardener/gardener/pull/8116#discussion_r1272081869 and revert this commit.

File: pkg/component/clusterautoscaler/cluster_autoscaler.go [hunk 19]
```diff
@@ -352,7 +352,8 @@ func (c *clusterAutoscaler) emptyManagedResource() *resourcesv1alpha1.ManagedRes
 }
 
 func (c *clusterAutoscaler) emptyManagedResourceSecret() *corev1.Secret {
-	return &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: managedresources.SecretName(managedResourceTargetName, true), Namespace: c.namespace}}
+	// TODO(dimityrmirchev): Remove this once mr secrets are turned into garbage-collectable, immutable secrets
```
Comment (shafeeqes) [inline pkg/component/clusterautoscaler/cluster_autoscaler.go]:
```suggestion
	// TODO(dimityrmirchev): Remove this once MR secrets are turned into garbage-collectable, immutable secrets
```
Can we add `v1.90` here too?

File: pkg/component/kubeapiserver/shoot_resources.go [hunk 20]
```diff
@@ -32,7 +32,8 @@ func (k *kubeAPIServer) emptyManagedResource() *resourcesv1alpha1.ManagedResourc
 }
 
 func (k *kubeAPIServer) emptyManagedResourceSecret() *corev1.Secret {
-	return &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: managedresources.SecretName(ManagedResourceName, true), Namespace: k.namespace}}
+	// TODO(dimityrmirchev): Remove this once mr secrets are turned into garbage-collectable, immutable secrets
```
Comment (shafeeqes) [inline pkg/component/kubeapiserver/shoot_resources.go]: Same here.

**Review by shafeeqes — COMMENTED** [review]
Thanks a lot for refactoring the tests ❤️

File: pkg/component/istio/istio_test.go (hunk 15 above)
Comment (dimityrmirchev) [reply pkg/component/istio/istio_test.go:2342]: This is because of https://github.com/gardener/gardener/pull/8116/commits/8dfded8bf092fd4704cbe6c6683d204f1ab3105d
See https://github.com/kubernetes/kubernetes/issues/118588 for the reasoning behind it.

**Comment by dimityrmirchev** [comment]
@shafeeqes thank you for reviewing! Regarding all the cleanup comments I plan to open an issue and assign myself if this PR gets merged so we do not forget and be able to track the process.

File: pkg/utils/managedresources/builder/managedresources.go [hunk 21]
```diff
@@ -199,7 +200,7 @@ func markSecretsAsGarbageCollectable(ctx context.Context, c client.Client, secre
 	return nil
 }
 
-func secretsFromRefs(obj *resourcesv1alpha1.ManagedResource, excludedNames map[string]struct{}) []*corev1.Secret {
+func secretsFromRefs[S sets.Set[string]](obj *resourcesv1alpha1.ManagedResource, excludedNames S) []*corev1.Secret {
 	secrets := make([]*corev1.Secret, 0, len(obj.Spec.SecretRefs))
 	for _, secretRef := range obj.Spec.SecretRefs {
 		if _, ok := excludedNames[secretRef.Name]; !ok {
```
Comment (shafeeqes) [inline pkg/utils/managedresources/builder/managedresources.go]: You can call `excludedNames.Has(secretRef.Name)` now.

File: pkg/utils/managedresources/builder/managedresources.go [hunk 22]
```diff
@@ -199,7 +200,7 @@ func markSecretsAsGarbageCollectable(ctx context.Context, c client.Client, secre
 	return nil
 }
 
-func secretsFromRefs(obj *resourcesv1alpha1.ManagedResource, excludedNames map[string]struct{}) []*corev1.Secret {
+func secretsFromRefs[S sets.Set[string]](obj *resourcesv1alpha1.ManagedResource, excludedNames S) []*corev1.Secret {
```
Comment (shafeeqes) [inline pkg/utils/managedresources/builder/managedresources.go]:
```suggestion
func secretsFromRefs(obj *resourcesv1alpha1.ManagedResource, excludedNames sets.Set[string]) []*corev1.Secret {
```

File: pkg/utils/managedresources/builder/managedresources.go (hunk 21 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/builder/managedresources.go]: I made the suggested changes, although the functionality stays unaffected and one could also argue that this does not improve readability over what the language already provides.

File: pkg/utils/managedresources/builder/managedresources.go [hunk 23]
```diff
@@ -112,28 +116,102 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
...
 func (m *ManagedResource) Delete(ctx context.Context) error {
 	return client.IgnoreNotFound(m.client.Delete(ctx, m.resource))
 }
+
+func markSecretsAsGarbageCollectable(ctx context.Context, c client.Client, secrets []*corev1.Secret) error {
+	for _, secret := range secrets {
+		if err := c.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+			return err
+		}
+		patch := client.StrategicMergeFrom(secret.DeepCopy())
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/managedresources.go:208]: You could exit early if the secret already has the garbage-collectable label to prevent empty paches.

File: charts/gardener/resource-manager/values.yaml [hunk 24]
```diff
@@ -76,6 +76,7 @@ global:
       garbageCollector:
         enabled: false
       # syncPeriod: 1h
+      # considerManagedResources: true
```
Comment (timuthy) [inline charts/gardener/resource-manager/values.yaml:79]: Please add this to `example/resource-manager/10-componentconfig.yaml`, as well.

File: pkg/utils/managedresources/managedresources.go (hunk 7 above)
Comment (timuthy) [reply pkg/utils/managedresources/managedresources.go:201]: Yep, that's the part I meant 👍

File: pkg/utils/managedresources/managedresources.go [hunk 25]
```diff
@@ -186,20 +189,43 @@ func deployManagedResource(ctx context.Context, secret *builder.Secret, managedR
...
+func Delete(ctx context.Context, c client.Client, namespace string, name string, secretNameWithPrefix bool) error {
 	secretName := SecretName(name, secretNameWithPrefix)
 
-	if err := builder.NewManagedResource(client).
-		WithNamespacedName(namespace, name).
-		Delete(ctx); err != nil {
-		return fmt.Errorf("could not delete managed resource '%s/%s': %w", namespace, name, err)
+	mr := &resourcesv1alpha1.ManagedResource{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace}}
+	mrKey := client.ObjectKeyFromObject(mr)
+	secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: namespace}}
```
Comment (timuthy) [inline pkg/utils/managedresources/managedresources.go]: Can we move this to the `if err != nil && apierrors.IsNotFound(err)` block?

File: pkg/utils/managedresources/managedresources.go [hunk 26]
```diff
@@ -186,20 +189,43 @@ func deployManagedResource(ctx context.Context, secret *builder.Secret, managedR
...
+	for _, secretRef := range mr.Spec.SecretRefs {
+		secretsToDelete = append(secretsToDelete, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+			Name:      secretRef.Name,
+			Namespace: namespace,
+		}})
+	}
+
+	// Delete the secrets first so we do not lose reference to them
+	// in case the mr gets deleted and something fails immediately after that.
+	// Finalizers should prevent the deletion of the secrets before the managed resource is deleted.
```
Comment (timuthy) [inline pkg/utils/managedresources/managedresources.go:229]: Nice 👍

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go [hunk 27]
```diff
@@ -326,25 +327,27 @@ func (r *Reconciler) delete(
...
+	timeoutSeedCtx, cancelCtx := context.WithTimeout(seedCtx, fiveSecond)
+	defer cancelCtx()
+	if err := managedresources.WaitUntilDeleted(timeoutSeedCtx, r.SeedClientSet.Client(), mr.Namespace, mr.Name); err != nil {
```
Comment (timuthy) [inline pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: Why do we need to wait here and don't stick to the re-enqueuing approach?

File: pkg/utils/managedresources/builder/managedresources.go (hunk 3 above)
Comment (timuthy) [reply pkg/utils/managedresources/builder/managedresources.go]: Thinking about it again, we can't simply remove it (even in a high version) since there is no guaranteed migration step that processes the MRs of all affected clusters. Should we add such a step in the startup proceedure of Gardenlet, to handle all MRs of class `seed`?

File: pkg/utils/managedresources/managedresources.go (hunk 25 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/managedresources.go]: I am not sure that we can. In the most recent version of this file the variables are used also outside the scope of the said `if` statement

File: pkg/utils/managedresources/managedresources.go (hunk 7 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/managedresources.go:201]: I will also remove the whole `Secrets` type, since it is not used anywhere in g/g and honestly I do not see it as any useful API. I will also include this in the release notes as a breaking change

File: pkg/utils/managedresources/builder/managedresources.go (hunk 23 above)
Comment (dimityrmirchev) [reply pkg/utils/managedresources/builder/managedresources.go:208]: Nice comment! Will do.

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go (hunk 27 above)
Comment (dimityrmirchev) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: My first thought was to use the high level function but now that I think more about it a better approach might be

1. Try getting the mr
2. if there are no errors - requeue
3. if err is not found - continue with the flow
4. if err is something else - we got different problem at hand (log and return the error)

Is this what you are suggesting?

Comment (timuthy) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: That's also fine. What I rather meant is that the current implementation of the delete procedure should be good enough and we probably shouldn't add a waiting step just to extract errors from the ManagedResource. Operators can still look at the ManagedResource in order to find out what's wrong.

Comment (dimityrmirchev) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: Hm, but imo we still need to check if the mr is successfully deleted before continuing with the namespace deletion, right?

**Comment by dimityrmirchev** [comment]
@timuthy For some reason I cannot continue the thread so I am quoting the question here

>Thinking about it again, we can't simply remove it (even in a high version) since there is no guaranteed migration step that processes the MRs of all affected clusters. Should we add such a step in the startup proceedure of Gardenlet, to handle all MRs of class seed?

In my opinion such migration steps are always troublesome and there are some cases where just leaving the backwards compatible code is the way to go. I think that this might be such case. WDYT about not removing this code at all and leaving it as it is. To me this might be the more pragmatic approach. It will also allow developers to even "adopt" a managed resource which was not previously managed with this library, i.e. they will get their old secrets garbage collected.

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go (hunk 27 above)
Comment (timuthy) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: Absolutely, and this already happens today (https://github.com/gardener/gardener/blob/ce96f9a15a81a2633f8d3ae64dccb1122460ec22/pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go#L330). So let me rephrase the quest: Why do we need to change something here in the scope of this PR? 😃

**Comment by timuthy** [comment]
> In my opinion such migration steps are always troublesome and there are some cases where just leaving the backwards compatible code is the way to go. I think that this might be such case. WDYT about not removing this code at all and leaving it as it is. To me this might be the more pragmatic approach. It will also allow developers to even "adopt" a managed resource which was not previously managed with this library, i.e. they will get their old secrets garbage collected.

Fine for me. Then let's remove the TODO and add a nice description why we need it the way it's implemented (delete old secrets first, then new ones) - I assume nobody will recall or understand this w/o any context later.

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go (hunk 27 above)
Comment (dimityrmirchev) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go]: Previously we were using `client.Delete()` which does not include all the cool logic behind `managedresources.Delete()`. So now when using `managedresources.Delete()` we still need to adapt the reconciler and ensure that the mr is actually gone before continuing.

I will make another commit with the suggested final proposal of this check

**Review by shafeeqes — COMMENTED** [review]
Thanks!
/lgtm

**Comment by timuthy** [comment]
Looks good so far, thanks for addressing the comments 🙂 I'm currently testing the final version of the PR in my local setup and then we can hopefully merge the changes (after rebase 😉)

File: pkg/utils/managedresources/builder/secrets.go [hunk 28]
```diff
@@ -47,9 +52,45 @@ func (s *Secret) WithNamespacedName(namespace, name string) *Secret {
...
+		s.secret.Labels[references.LabelKeyGarbageCollectable] = references.LabelValueGarbageCollectable
+		for k, v := range labels {
+			s.secret.Labels[k] = v
+		}
+		return s
+	}
+
+	s.secret.Labels = cloneLabels(labels)
+	return s
+}
```
Comment (timuthy) [inline pkg/utils/managedresources/builder/secrets.go]:
```suggestion
// WithLabels sets the labels. The label "resources.gardener.cloud/garbage-collectable-reference" is retained
// if it already exists in the current labels.
func (s *Secret) WithLabels(labels map[string]string) *Secret {
	if s.secret.Labels == nil {
		s.secret.Labels = utils.MergeStringMaps(labels)
		return s
	}

	_, ok := s.secret.Labels[references.LabelKeyGarbageCollectable]
	if ok && pointer.BoolDeref(s.secret.Immutable, false) {
		s.secret.Labels = map[string]string{references.LabelKeyGarbageCollectable: references.LabelValueGarbageCollectable}
	}

	s.secret.Labels = utils.MergeStringMaps(labels, s.secret.Labels)
	return s
}
```

We already use `utils.MergeStringMaps` to copy maps in other places.

File: example/resource-manager/10-componentconfig.yaml [hunk 29]
```diff
@@ -40,6 +40,7 @@ controllers:
   garbageCollector:
     enabled: true
     syncPeriod: 1h
+    # considerManagedResources: true
```
Comment (timuthy) [inline example/resource-manager/10-componentconfig.yaml]:
```suggestion
  # considerManagedResources: true
```

File: pkg/resourcemanager/apis/config/types.go [hunk 30]
```diff
@@ -144,6 +144,9 @@ type GarbageCollectorControllerConfig struct {
 	Enabled bool
 	// SyncPeriod is the duration how often the controller performs its reconciliation.
 	SyncPeriod *metav1.Duration
+	// ConsiderManagedResources specifies if managed resources should be considered
+	// as owners of a garbage collectable secret.
```
Comment (timuthy) [inline pkg/resourcemanager/apis/config/types.go]:
```suggestion
	// ConsiderManagedResources specifies if managed resources should be considered
	// as owners of garbage collectable secrets.
```

File: pkg/resourcemanager/apis/config/v1alpha1/types.go [hunk 31]
```diff
@@ -158,6 +158,10 @@ type GarbageCollectorControllerConfig struct {
 	// SyncPeriod is the duration how often the controller performs its reconciliation.
 	// +optional
 	SyncPeriod *metav1.Duration `json:"syncPeriod,omitempty"`
+	// ConsiderManagedResources specifies if managed resources should be considered
+	// as owners of a garbage collectable secret.
```
Comment (timuthy) [inline pkg/resourcemanager/apis/config/v1alpha1/types.go]:
```suggestion
	// ConsiderManagedResources specifies if managed resources should be considered
	// as owners of garbage collectable secrets.
```

File: pkg/resourcemanager/controller/garbagecollector/references/references.go [hunk 32]
```diff
@@ -147,13 +151,29 @@ func InjectAnnotations(obj runtime.Object, additional ...string) error {
...
+		o.Annotations = mergeAnnotations(o.Annotations, referenceAnnotations)
+
 	default:
 		return fmt.Errorf("unhandled object type %T", obj)
 	}
 
 	return nil
 }
 
+func computeAnnotationsFromLocalObjRefs(refs []corev1.LocalObjectReference, kind string, additional ...string) map[string]string {
```
Comment (timuthy) [inline pkg/resourcemanager/controller/garbagecollector/references/references.go:165]:
```suggestion
func computeAnnotationsFromLocalObjRefs(refs []corev1.LocalObjectReference, additional ...string) map[string]string {
```

File: pkg/resourcemanager/controller/garbagecollector/reconciler.go [hunk 33]
```diff
@@ -99,6 +101,10 @@ func (r *Reconciler) Reconcile(reconcileCtx context.Context, _ reconcile.Request
 		}
 	)
 
+	if pointer.BoolDeref(r.Config.ConsiderManagedResources, false) {
```
Comment (timuthy) [inline pkg/resourcemanager/controller/garbagecollector/reconciler.go:104]: Can you please enhance the GC integration test in `test/integration/resourcemanager/garbagecollector/garbagecollector_test.go`?

**Review by timuthy — COMMENTED** [review]
Local testing went well 👏 
Some last suggestions from my side, then we're good to go.

File: pkg/resourcemanager/controller/garbagecollector/reconciler.go (hunk 33 above)
Comment (dimityrmirchev) [reply pkg/resourcemanager/controller/garbagecollector/reconciler.go:104]: Done

**Comment by dimityrmirchev** [comment]
@timuthy I did some final changes, as well as addressed the comments. Please review the commits from https://github.com/gardener/gardener/pull/8116/commits/1fdbdb93fe74657c6858b3606b5cfdaa6968a244 onwards (the changes are not that much, mainly some small improvements). I also hope that tests will not flake anymore with the latest changes (I was not able to verify this because for some reason I am having troubles running [the debugging integration tests guide](https://github.com/gardener/gardener/blob/master/docs/development/testing.md#debugging-integration-tests) on my dev machine)

**Review by timuthy — COMMENTED** [review]
Nice, thank you 👏 
/lgtm

**Review by shafeeqes — COMMENTED** [review]
/lgtm

**Review by timuthy — COMMENTED** [review]
/approve
